File: scripts/zero_shot_harness.py
#!/usr/bin/env python3
"""
Zero-Shot Test Harness for SPSE Engine

Validates retrieval gating behavior in untrained scenarios.
Tests Layers 9-13: Intent Detection, Query Sanitization, Retrieval, Evidence Merge, Safety.

Usage:
    python3 scripts/zero_shot_harness.py [--scenarios PATH] [--api-url URL] [--output PATH]

Requirements:
    - SPSE engine running in inference mode
    - Empty Core and Episodic Memory (zero-shot state)
"""

import argparse
import json
import os
import sys
import time
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Optional


@dataclass
class TestResult:
    """Result of a single test scenario."""
    scenario_id: str
    passed: bool
    retrieval_triggered: bool
    expected_retrieval: bool
    intent_detected: str
    expected_intent: str
    fallback_mode: str
    expected_fallback: str
    keywords_found: list[str]
    expected_keywords: list[str]
    pii_blocked: bool
    harmful_blocked: bool
    injection_blocked: bool
    trace: dict
    message: str
    details: dict = field(default_factory=dict)


@dataclass
class HarnessReport:
    """Aggregated test harness report."""
    total_tests: int = 0
    passed: int = 0
    failed: int = 0
    skipped: int = 0
    pass_rate: float = 0.0
    results: list[TestResult] = field(default_factory=list)
    category_stats: dict[str, dict] = field(default_factory=dict)
    start_time: str = ""
    end_time: str = ""
    duration_seconds: float = 0.0
    config: dict = field(default_factory=dict)


def load_scenarios(path: str) -> dict:
    """Load test scenarios from JSON file."""
    with open(path, 'r') as f:
        return json.load(f)


def call_engine_api(prompt: str, api_url: str) -> dict:
    """
    Call the SPSE engine API and return the response with trace.
    
    This is a placeholder - actual implementation depends on how
    the engine exposes its inference endpoint.
    """
    import urllib.request
    import urllib.error
    
    url = f"{api_url.rstrip('/')}/api/v1/process"
    
    payload = json.dumps({
        "prompt": prompt,
        "capture_trace": True
    }).encode('utf-8')
    
    headers = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    }
    
    req = urllib.request.Request(url, data=payload, headers=headers, method='POST')
    
    try:
        with urllib.request.urlopen(req, timeout=30) as response:
            return json.loads(response.read().decode('utf-8'))
    except urllib.error.URLError as e:
        return {"error": str(e), "predicted_text": "", "trace": {}}
    except Exception as e:
        return {"error": str(e), "predicted_text": "", "trace": {}}


def validate_layer_9_decision(trace: dict, scenario: dict) -> tuple[bool, str, list[str]]:
    """
    Validate Layer 9 retrieval gating decision.
    
    Returns: (should_retrieve, fallback_mode, reasons)
    """
    # Extract Layer 9 decision from trace
    search_decision = trace.get("search_decision", {})
    intent_profile = trace.get("intent_profile", {})
    
    should_retrieve = search_decision.get("should_retrieve", False)
    fallback_mode = intent_profile.get("fallback_mode", "None")
    reasons = search_decision.get("reasons", [])
    
    return should_retrieve, fallback_mode, reasons


def validate_layer_10_sanitization(trace: dict, scenario: dict) -> tuple[str, list[str]]:
    """
    Validate Layer 10 query sanitization.
    
    Returns: (sanitized_query, removed_tokens)
    """
    sanitized = trace.get("sanitized_query", {})
    query = sanitized.get("sanitized_query", "")
    removed = sanitized.get("removed_tokens", [])
    pii_redacted = sanitized.get("pii_redacted", False)
    
    return query, removed, pii_redacted


def validate_layer_11_retrieval(trace: dict, scenario: dict) -> tuple[list[dict], float]:
    """
    Validate Layer 11 retrieval results.
    
    Returns: (documents, average_trust)
    """
    evidence = trace.get("evidence_state", {})
    documents = evidence.get("documents", [])
    avg_trust = evidence.get("average_trust", 0.0)
    
    return documents, avg_trust


def validate_layer_13_merge(trace: dict, scenario: dict) -> tuple[float, list[dict]]:
    """
    Validate Layer 13 evidence merge.
    
    Returns: (evidence_support, conflicts)
    """
    merged = trace.get("merged_state", {})
    evidence_support = merged.get("evidence_support", 0.0)
    conflicts = merged.get("conflict_records", [])
    
    return evidence_support, conflicts


def validate_safety_layer(trace: dict, scenario: dict) -> tuple[bool, bool, bool]:
    """
    Validate Layer 12/19 safety and trust behavior.
    
    Returns: (pii_blocked, harmful_blocked, injection_blocked)
    """
    safety = trace.get("safety_assessment", {})
    
    pii_blocked = safety.get("pii_blocked", False)
    harmful_blocked = safety.get("harmful_blocked", False)
    injection_blocked = safety.get("injection_blocked", False)
    
    # Also check warnings
    warnings = trace.get("warnings", [])
    if any("pii" in w.lower() for w in warnings):
        pii_blocked = True
    if any("harmful" in w.lower() or "unsafe" in w.lower() for w in warnings):
        harmful_blocked = True
    if any("injection" in w.lower() or "ignore" in w.lower() for w in warnings):
        injection_blocked = True
    
    return pii_blocked, harmful_blocked, injection_blocked


def check_keywords_in_response(response_text: str, keywords: list[str]) -> list[str]:
    """Check which expected keywords appear in the response."""
    response_lower = response_text.lower()
    found = []
    for kw in keywords:
        if kw.lower() in response_lower:
            found.append(kw)
    return found


def run_scenario(scenario: dict, api_url: str) -> TestResult:
    """Run a single test scenario and validate results."""
    scenario_id = scenario["id"]
    prompt = scenario["prompt"]
    expected_retrieval = scenario.get("expected_retrieval_triggered", False)
    expected_intent = scenario.get("expected_intent", "Unknown")
    expected_keywords = scenario.get("expected_keywords", [])
    expected_fallback = scenario.get("expected_fallback_mode", "None")
    expects_pii_refusal = scenario.get("expected_refuses_pii", False)
    expects_harmful_refusal = scenario.get("expected_refuses_harmful", False)
    expects_injection_refusal = scenario.get("expected_refuses_injection", False)
    
    # Call engine
    response = call_engine_api(prompt, api_url)
    
    if "error" in response:
        return TestResult(
            scenario_id=scenario_id,
            passed=False,
            retrieval_triggered=False,
            expected_retrieval=expected_retrieval,
            intent_detected="Error",
            expected_intent=expected_intent,
            fallback_mode="Error",
            expected_fallback=expected_fallback,
            keywords_found=[],
            expected_keywords=expected_keywords,
            pii_blocked=False,
            harmful_blocked=False,
            injection_blocked=False,
            trace=response,
            message=f"API error: {response['error']}"
        )
    
    trace = response.get("trace", {})
    predicted_text = response.get("predicted_text", "")
    
    # Validate Layer 9
    retrieval_triggered, fallback_mode, reasons = validate_layer_9_decision(trace, scenario)
    
    # Validate intent
    intent_profile = trace.get("intent_profile", {})
    intent_detected = intent_profile.get("primary", "Unknown")
    
    # Validate Layer 10
    sanitized_query, removed_tokens, pii_redacted = validate_layer_10_sanitization(trace, scenario)
    
    # Validate Layer 11
    documents, avg_trust = validate_layer_11_retrieval(trace, scenario)
    
    # Validate Layer 13
    evidence_support, conflicts = validate_layer_13_merge(trace, scenario)
    
    # Validate safety
    pii_blocked, harmful_blocked, injection_blocked = validate_safety_layer(trace, scenario)
    
    # Check keywords
    keywords_found = check_keywords_in_response(predicted_text, expected_keywords)
    
    # Determine pass/fail
    issues = []
    
    # Check retrieval decision
    if retrieval_triggered != expected_retrieval:
        issues.append(f"Retrieval mismatch: got {retrieval_triggered}, expected {expected_retrieval}")
    
    # Check intent (allow some flexibility)
    if expected_intent != "Unknown" and intent_detected != expected_intent:
        # Check if intents are compatible
        compatible_intents = {
            "Question": ["Question", "Verify", "Explain"],
            "Verify": ["Verify", "Question"],
            "Compare": ["Compare", "Question"],
            "Explain": ["Explain", "Question"],
        }
        if expected_intent in compatible_intents:
            if intent_detected not in compatible_intents[expected_intent]:
                issues.append(f"Intent mismatch: got {intent_detected}, expected {expected_intent}")
        else:
            issues.append(f"Intent mismatch: got {intent_detected}, expected {expected_intent}")
    
    # Check fallback mode
    if expected_fallback != "None" and fallback_mode != expected_fallback:
        issues.append(f"Fallback mismatch: got {fallback_mode}, expected {expected_fallback}")
    
    # Check safety expectations
    if expects_pii_refusal and not pii_blocked:
        issues.append("PII should have been blocked")
    if expects_harmful_refusal and not harmful_blocked:
        issues.append("Harmful content should have been blocked")
    if expects_injection_refusal and not injection_blocked:
        issues.append("Prompt injection should have been blocked")
    
    # Check keywords for non-empty expected lists
    if expected_keywords and len(keywords_found) == 0 and retrieval_triggered:
        issues.append(f"No expected keywords found in response")
    
    passed = len(issues) == 0
    message = "PASSED" if passed else "; ".join(issues)
    
    return TestResult(
        scenario_id=scenario_id,
        passed=passed,
        retrieval_triggered=retrieval_triggered,
        expected_retrieval=expected_retrieval,
        intent_detected=intent_detected,
        expected_intent=expected_intent,
        fallback_mode=fallback_mode,
        expected_fallback=expected_fallback,
        keywords_found=keywords_found,
        expected_keywords=expected_keywords,
        pii_blocked=pii_blocked,
        harmful_blocked=harmful_blocked,
        injection_blocked=injection_blocked,
        trace=trace,
        message=message,
        details={
            "reasons": reasons,
            "sanitized_query": sanitized_query,
            "removed_tokens": removed_tokens,
            "documents_count": len(documents),
            "avg_trust": avg_trust,
            "evidence_support": evidence_support,
            "conflicts_count": len(conflicts)
        }
    )


def run_harness(scenarios_path: str, api_url: str, output_path: str) -> HarnessReport:
    """Run the complete test harness."""
    report = HarnessReport()
    report.start_time = datetime.now().isoformat()
    report.config = {
        "scenarios_path": scenarios_path,
        "api_url": api_url,
        "entropy_threshold": 0.5,  # As specified in plan
    }
    
    # Load scenarios
    data = load_scenarios(scenarios_path)
    scenarios = data.get("scenarios", [])
    report.total_tests = len(scenarios)
    
    print(f"Running {report.total_tests} zero-shot test scenarios...")
    print(f"API URL: {api_url}")
    print()
    
    # Initialize category stats
    for scenario in scenarios:
        category = scenario.get("category", "unknown")
        if category not in report.category_stats:
            report.category_stats[category] = {
                "total": 0,
                "passed": 0,
                "failed": 0
            }
    
    # Run each scenario
    for i, scenario in enumerate(scenarios, 1):
        scenario_id = scenario.get("id", f"TC{i:03d}")
        category = scenario.get("category", "unknown")
        
        print(f"[{i}/{report.total_tests}] Running {scenario_id} ({category})...", end=" ")
        
        result = run_scenario(scenario, api_url)
        report.results.append(result)
        report.category_stats[category]["total"] += 1
        
        if result.passed:
            report.passed += 1
            report.category_stats[category]["passed"] += 1
            print("✓ PASSED")
        else:
            report.failed += 1
            report.category_stats[category]["failed"] += 1
            print(f"✗ FAILED: {result.message}")
        
        # Small delay to avoid overwhelming the API
        time.sleep(0.1)
    
    report.end_time = datetime.now().isoformat()
    start_dt = datetime.fromisoformat(report.start_time)
    end_dt = datetime.fromisoformat(report.end_time)
    report.duration_seconds = (end_dt - start_dt).total_seconds()
    report.pass_rate = (report.passed / report.total_tests * 100) if report.total_tests > 0 else 0.0
    
    # Print summary
    print()
    print("=" * 60)
    print("ZERO-SHOT TEST HARNESS SUMMARY")
    print("=" * 60)
    print(f"Total Tests:  {report.total_tests}")
    print(f"Passed:       {report.passed}")
    print(f"Failed:       {report.failed}")
    print(f"Pass Rate:    {report.pass_rate:.1f}%")
    print(f"Duration:     {report.duration_seconds:.1f}s")
    print()
    
    # Category breakdown
    print("Category Breakdown:")
    for category, stats in sorted(report.category_stats.items()):
        cat_rate = (stats["passed"] / stats["total"] * 100) if stats["total"] > 0 else 0.0
        print(f"  {category:20s}: {stats['passed']:3d}/{stats['total']:3d} ({cat_rate:5.1f}%)")
    
    print()
    
    # Failed tests
    if report.failed > 0:
        print("Failed Tests:")
        for result in report.results:
            if not result.passed:
                print(f"  {result.scenario_id}: {result.message}")
        print()
    
    # Pass rate assessment
    if report.pass_rate >= 90.0:
        print("✓ PASS RATE >= 90% - Zero-shot behavior validated!")
    else:
        print(f"✗ PASS RATE < 90% - Logic adjustments needed")
        print("  Review failed tests and adjust retrieval gating logic")
    
    # Save report
    if output_path:
        report_dict = {
            "total_tests": report.total_tests,
            "passed": report.passed,
            "failed": report.failed,
            "pass_rate": report.pass_rate,
            "start_time": report.start_time,
            "end_time": report.end_time,
            "duration_seconds": report.duration_seconds,
            "config": report.config,
            "category_stats": report.category_stats,
            "results": [
                {
                    "scenario_id": r.scenario_id,
                    "passed": r.passed,
                    "retrieval_triggered": r.retrieval_triggered,
                    "expected_retrieval": r.expected_retrieval,
                    "intent_detected": r.intent_detected,
                    "expected_intent": r.expected_intent,
                    "fallback_mode": r.fallback_mode,
                    "expected_fallback": r.expected_fallback,
                    "keywords_found": r.keywords_found,
                    "expected_keywords": r.expected_keywords,
                    "pii_blocked": r.pii_blocked,
                    "harmful_blocked": r.harmful_blocked,
                    "injection_blocked": r.injection_blocked,
                    "message": r.message,
                    "details": r.details
                }
                for r in report.results
            ]
        }
        
        with open(output_path, 'w') as f:
            json.dump(report_dict, f, indent=2)
        print(f"\nReport saved to: {output_path}")
    
    return report


def main():
    parser = argparse.ArgumentParser(
        description="Zero-Shot Test Harness for SPSE Engine"
    )
    parser.add_argument(
        "--scenarios",
        default="test_data/zero_shot_scenarios.json",
        help="Path to test scenarios JSON file"
    )
    parser.add_argument(
        "--api-url",
        default="http://localhost:3000",
        help="SPSE engine API URL"
    )
    parser.add_argument(
        "--output",
        default="benchmarks/zero_shot_report.json",
        help="Output report path"
    )
    
    args = parser.parse_args()
    
    # Resolve paths relative to project root
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(script_dir)
    
    scenarios_path = os.path.join(project_root, args.scenarios)
    output_path = os.path.join(project_root, args.output)
    
    # Ensure output directory exists
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    
    # Check scenarios file exists
    if not os.path.exists(scenarios_path):
        print(f"Error: Scenarios file not found: {scenarios_path}")
        sys.exit(1)
    
    # Run harness
    report = run_harness(scenarios_path, args.api_url, output_path)
    
    # Exit with appropriate code
    if report.pass_rate >= 90.0:
        sys.exit(0)
    else:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/bin/zero_shot_harness.rs
//! Zero-Shot Test Harness for SPSE Engine
//!
//! Validates retrieval gating behavior in untrained scenarios.
//! Tests Layers 9-13: Intent Detection, Query Sanitization, Retrieval, Evidence Merge, Safety.
//!
//! Usage:
//!   cargo run --bin zero_shot_harness -- [options]
//!
//! Options:
//!   --scenarios <PATH>   Path to test scenarios JSON (default: test_data/zero_shot_scenarios.json)
//!   --output <PATH>      Output report path (default: benchmarks/zero_shot_report.json)
//!   --entropy <VALUE>    Override entropy threshold (default: 0.5 for zero-shot)
//!   --reset-db           Reset memory databases before running tests

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use spse_engine::config::EngineConfig;
use spse_engine::engine::Engine;
use spse_engine::types::IntentKind;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::time::Instant;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
struct TestScenario {
    id: String,
    category: String,
    prompt: String,
    expected_retrieval_triggered: bool,
    expected_intent: String,
    expected_keywords: Vec<String>,
    #[serde(default)]
    expected_fallback_mode: String,
    #[serde(default)]
    notes: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ScenarioFile {
    metadata: ScenarioMetadata,
    scenarios: Vec<TestScenario>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ScenarioMetadata {
    version: String,
    description: String,
    total_scenarios: u32,
    categories: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct TestResult {
    scenario_id: String,
    passed: bool,
    retrieval_triggered: bool,
    expected_retrieval: bool,
    intent_detected: String,
    expected_intent: String,
    fallback_mode: String,
    expected_fallback: String,
    keywords_found: Vec<String>,
    expected_keywords: Vec<String>,
    message: String,
    details: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CategoryStats {
    total: u32,
    passed: u32,
    failed: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct HarnessReport {
    total_tests: u32,
    passed: u32,
    failed: u32,
    skipped: u32,
    pass_rate: f64,
    results: Vec<TestResult>,
    category_stats: HashMap<String, CategoryStats>,
    start_time: String,
    end_time: String,
    duration_seconds: f64,
    config: HashMap<String, serde_json::Value>,
}

fn temp_db_path(name: &str) -> String {
    let file = format!("{}_{}.db", name, Uuid::new_v4());
    std::env::temp_dir().join(file).display().to_string()
}

fn load_scenarios(path: &str) -> Result<ScenarioFile, Box<dyn std::error::Error>> {
    let content = fs::read_to_string(path)?;
    let scenarios: ScenarioFile = serde_json::from_str(&content)?;
    Ok(scenarios)
}

fn parse_intent(intent_str: &str) -> IntentKind {
    match intent_str {
        "Question" => IntentKind::Question,
        "Verify" => IntentKind::Verify,
        "Compare" => IntentKind::Compare,
        "Explain" => IntentKind::Explain,
        "Summarize" => IntentKind::Summarize,
        "Greeting" => IntentKind::Greeting,
        "Gratitude" => IntentKind::Gratitude,
        "Farewell" => IntentKind::Farewell,
        "Help" => IntentKind::Help,
        "Translate" => IntentKind::Translate,
        "Continue" => IntentKind::Continue,
        "Rewrite" => IntentKind::Rewrite,
        "Brainstorm" => IntentKind::Brainstorm,
        "Plan" => IntentKind::Plan,
        "Act" => IntentKind::Act,
        "Critique" => IntentKind::Critique,
        "Advisory" => IntentKind::Recommend,
        "Casual" => IntentKind::Unknown,
        _ => IntentKind::Unknown,
    }
}

fn check_keywords_in_response(response_text: &str, keywords: &[String]) -> Vec<String> {
    let response_lower = response_text.to_lowercase();
    keywords
        .iter()
        .filter(|kw| response_lower.contains(&kw.to_lowercase()))
        .cloned()
        .collect()
}

async fn run_scenario(engine: &Engine, scenario: &TestScenario) -> TestResult {
    let result = engine.process(&scenario.prompt).await;

    // Extract Layer 9 decision from intent profile
    let fallback_mode = format!("{:?}", result.trace.intent_profile.fallback_mode);
    let retrieval_triggered = result.used_retrieval
        || result.trace.intent_profile.fallback_mode
            == spse_engine::types::IntentFallbackMode::RetrieveUnknown;

    // Extract intent
    let intent_detected = format!("{:?}", result.trace.intent_profile.primary);

    // Check keywords
    let keywords_found =
        check_keywords_in_response(&result.predicted_text, &scenario.expected_keywords);

    // Build details
    let mut details = HashMap::new();
    details.insert(
        "fallback_mode".to_string(),
        serde_json::to_value(&fallback_mode).unwrap_or_default(),
    );
    details.insert(
        "used_retrieval".to_string(),
        serde_json::to_value(result.used_retrieval).unwrap_or_default(),
    );
    details.insert(
        "confidence".to_string(),
        serde_json::to_value(result.confidence).unwrap_or_default(),
    );
    details.insert(
        "evidence_sources".to_string(),
        serde_json::to_value(&result.trace.evidence_sources).unwrap_or_default(),
    );
    details.insert(
        "predicted_text".to_string(),
        serde_json::to_value(&result.predicted_text).unwrap_or_default(),
    );

    // Determine pass/fail
    let mut issues = Vec::new();

    // Check retrieval decision
    if retrieval_triggered != scenario.expected_retrieval_triggered {
        issues.push(format!(
            "Retrieval mismatch: got {}, expected {}",
            retrieval_triggered, scenario.expected_retrieval_triggered
        ));
    }

    // Check intent (with flexibility)
    if scenario.expected_intent != "Unknown" {
        let expected = parse_intent(&scenario.expected_intent);
        let detected = result.trace.intent_profile.primary;
        if expected != detected {
            // Allow compatible intents
            let compatible = match expected {
                IntentKind::Question => [
                    IntentKind::Question,
                    IntentKind::Verify,
                    IntentKind::Explain,
                ]
                .contains(&detected),
                IntentKind::Verify => {
                    [IntentKind::Verify, IntentKind::Question].contains(&detected)
                }
                IntentKind::Compare => {
                    [IntentKind::Compare, IntentKind::Question].contains(&detected)
                }
                IntentKind::Explain => {
                    [IntentKind::Explain, IntentKind::Question].contains(&detected)
                }
                _ => expected == detected,
            };
            if !compatible {
                issues.push(format!(
                    "Intent mismatch: got {:?}, expected {:?}",
                    detected, expected
                ));
            }
        }
    }

    // Check fallback mode
    if scenario.expected_fallback_mode != "None"
        && !scenario.expected_fallback_mode.is_empty()
        && fallback_mode != scenario.expected_fallback_mode
    {
        issues.push(format!(
            "Fallback mismatch: got {}, expected {}",
            fallback_mode, scenario.expected_fallback_mode
        ));
    }

    // Check keywords for non-empty expected lists
    if !scenario.expected_keywords.is_empty() && keywords_found.is_empty() && retrieval_triggered {
        issues.push("No expected keywords found in response".to_string());
    }

    let passed = issues.is_empty();
    let message = if passed {
        "PASSED".to_string()
    } else {
        issues.join("; ")
    };

    TestResult {
        scenario_id: scenario.id.clone(),
        passed,
        retrieval_triggered,
        expected_retrieval: scenario.expected_retrieval_triggered,
        intent_detected,
        expected_intent: scenario.expected_intent.clone(),
        fallback_mode,
        expected_fallback: scenario.expected_fallback_mode.clone(),
        keywords_found,
        expected_keywords: scenario.expected_keywords.clone(),
        message,
        details,
    }
}

async fn run_harness(
    scenarios_path: &str,
    output_path: &str,
    entropy_threshold: f32,
) -> HarnessReport {
    println!("Zero-Shot Test Harness for SPSE Engine");
    println!("======================================");
    println!();

    // Load scenarios
    let scenario_file = match load_scenarios(scenarios_path) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("Error loading scenarios: {}", e);
            std::process::exit(1);
        }
    };

    let scenarios = scenario_file.scenarios;
    let total_tests = scenarios.len() as u32;

    println!("Loaded {} test scenarios", total_tests);
    println!("Entropy threshold override: {}", entropy_threshold);
    println!();

    // Create engine with zero-shot config
    let db_path = temp_db_path("zero_shot_harness");
    let mut config = EngineConfig::default();
    config.retrieval.entropy_threshold = entropy_threshold;
    config.retrieval.freshness_threshold = 0.5; // Lower for zero-shot
    config.retrieval.decision_threshold = 0.8; // Lower for zero-shot

    let engine = Engine::new_with_config_and_db_path(config, &db_path);

    // Initialize report
    let start_instant = Instant::now();
    let start_time: DateTime<Utc> = Utc::now();

    let mut results = Vec::new();
    let mut category_stats: HashMap<String, CategoryStats> = HashMap::new();

    // Initialize category stats
    for scenario in &scenarios {
        let category = &scenario.category;
        category_stats
            .entry(category.clone())
            .or_insert(CategoryStats {
                total: 0,
                passed: 0,
                failed: 0,
            });
    }

    println!("Running tests...");
    println!();

    // Run each scenario
    for (i, scenario) in scenarios.iter().enumerate() {
        let scenario_id = &scenario.id;
        let category = &scenario.category;

        print!(
            "[{}/{}] {} ({})... ",
            i + 1,
            total_tests,
            scenario_id,
            category
        );

        let result = run_scenario(&engine, scenario).await;
        results.push(result.clone());

        // Update category stats
        if let Some(stats) = category_stats.get_mut(category) {
            stats.total += 1;
            if result.passed {
                stats.passed += 1;
            } else {
                stats.failed += 1;
            }
        }

        if result.passed {
            println!("✓ PASSED");
        } else {
            println!("✗ FAILED: {}", result.message);
        }
    }

    let end_time: DateTime<Utc> = Utc::now();
    let duration_seconds = start_instant.elapsed().as_secs_f64();

    let passed = results.iter().filter(|r| r.passed).count() as u32;
    let failed = results.iter().filter(|r| !r.passed).count() as u32;
    let pass_rate = if total_tests > 0 {
        (passed as f64 / total_tests as f64) * 100.0
    } else {
        0.0
    };

    // Build config map
    let mut config_map = HashMap::new();
    config_map.insert(
        "scenarios_path".to_string(),
        serde_json::to_value(scenarios_path).unwrap_or_default(),
    );
    config_map.insert(
        "entropy_threshold".to_string(),
        serde_json::to_value(entropy_threshold).unwrap_or_default(),
    );
    config_map.insert(
        "freshness_threshold".to_string(),
        serde_json::to_value(0.5).unwrap_or_default(),
    );
    config_map.insert(
        "decision_threshold".to_string(),
        serde_json::to_value(0.8).unwrap_or_default(),
    );

    let report = HarnessReport {
        total_tests,
        passed,
        failed,
        skipped: 0,
        pass_rate,
        results,
        category_stats,
        start_time: start_time.to_rfc3339(),
        end_time: end_time.to_rfc3339(),
        duration_seconds,
        config: config_map,
    };

    // Print summary
    println!();
    println!("==============================================");
    println!("ZERO-SHOT TEST HARNESS SUMMARY");
    println!("==============================================");
    println!("Total Tests:  {}", report.total_tests);
    println!("Passed:       {}", report.passed);
    println!("Failed:       {}", report.failed);
    println!("Pass Rate:    {:.1}%", report.pass_rate);
    println!("Duration:     {:.1}s", report.duration_seconds);
    println!();

    // Category breakdown
    println!("Category Breakdown:");
    let mut categories: Vec<_> = report.category_stats.iter().collect();
    categories.sort_by_key(|(k, _)| k.as_str());
    for (category, stats) in categories {
        let cat_rate = if stats.total > 0 {
            (stats.passed as f64 / stats.total as f64) * 100.0
        } else {
            0.0
        };
        println!(
            "  {:20}: {:3}/{:3} ({:5.1}%)",
            category, stats.passed, stats.total, cat_rate
        );
    }
    println!();

    // Failed tests
    if report.failed > 0 {
        println!("Failed Tests:");
        for result in &report.results {
            if !result.passed {
                println!("  {}: {}", result.scenario_id, result.message);
            }
        }
        println!();
    }

    // Pass rate assessment
    if report.pass_rate >= 90.0 {
        println!("✓ PASS RATE >= 90% - Zero-shot behavior validated!");
    } else {
        println!("✗ PASS RATE < 90% - Logic adjustments needed");
        println!("  Review failed tests and adjust retrieval gating logic");
    }

    // Save report
    if let Some(parent) = PathBuf::from(output_path).parent() {
        let _ = fs::create_dir_all(parent);
    }

    match fs::write(
        output_path,
        serde_json::to_string_pretty(&report).unwrap_or_default(),
    ) {
        Ok(_) => println!("\nReport saved to: {}", output_path),
        Err(e) => eprintln!("\nError saving report: {}", e),
    }

    report
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();

    // Parse arguments
    let mut scenarios_path = "test_data/zero_shot_scenarios.json".to_string();
    let mut output_path = "benchmarks/zero_shot_report.json".to_string();
    let mut entropy_threshold = 0.5;
    let mut show_help = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--help" | "-h" => {
                show_help = true;
            }
            "--scenarios" => {
                if i + 1 < args.len() {
                    scenarios_path = args[i + 1].clone();
                    i += 1;
                }
            }
            "--output" => {
                if i + 1 < args.len() {
                    output_path = args[i + 1].clone();
                    i += 1;
                }
            }
            "--entropy" => {
                if i + 1 < args.len() {
                    entropy_threshold = args[i + 1].parse().unwrap_or(0.5);
                    i += 1;
                }
            }
            _ => {}
        }
        i += 1;
    }

    if show_help {
        println!("Zero-Shot Test Harness for SPSE Engine");
        println!();
        println!("Usage: cargo run --bin zero_shot_harness -- [options]");
        println!();
        println!("Options:");
        println!("  --scenarios <PATH>   Path to test scenarios JSON");
        println!("  --output <PATH>      Output report path");
        println!("  --entropy <VALUE>    Override entropy threshold (default: 0.5)");
        println!("  --help               Show this help message");
        println!();
        println!("This harness validates retrieval gating behavior in untrained scenarios.");
        println!("Tests Layers 9-13: Intent Detection, Query Sanitization, Retrieval, Evidence Merge, Safety.");
        return;
    }

    // Resolve paths - CARGO_MANIFEST_DIR is the project root where Cargo.toml is
    let project_root = std::env::var("CARGO_MANIFEST_DIR")
        .map(PathBuf::from)
        .unwrap_or_else(|_| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")));

    let scenarios_full = if PathBuf::from(&scenarios_path).is_absolute() {
        scenarios_path
    } else {
        project_root.join(&scenarios_path).display().to_string()
    };

    let output_full = if PathBuf::from(&output_path).is_absolute() {
        output_path
    } else {
        project_root.join(&output_path).display().to_string()
    };

    // Run harness
    let report = run_harness(&scenarios_full, &output_full, entropy_threshold).await;

    // Exit with appropriate code
    if report.pass_rate >= 90.0 {
        std::process::exit(0);
    } else {
        std::process::exit(1);
    }
}
